code/modules/antagonists/infection/infection/infections/upgrades.dm

```diff
@@ -0,0 +1,398 @@
+/datum/component/infection/upgrade
+	// display stuff
+	var/name = ""
+	var/description = ""
+	var/radial_icon = 'icons/mob/blob.dmi'
+	var/radial_icon_state = ""
+
+	// application
+	var/cost = 0
+	var/increasing_cost = 0 // the amount the cost increases every time the upgrade is purchased
+	var/times = 1 // times the upgrade can be bought
+	var/bought = 0 // how many times the upgrade has been bought
+
+/datum/component/infection/upgrade/Initialize()
+	. = ..()
+	if(bought)
+		register_signals()
+
+/datum/component/infection/upgrade/proc/register_signals()
+	RegisterSignal(parent, COMSIG_INFECTION_LIFE_TICK, .proc/life_tick)
```

Comment:
If something needs to process it should not piggyback off another thing processing. Add it to a subsystem on its own if it needs to process.